clash-lib/src/Clash/Primitives/Scaffold.hs

```diff
@@ -0,0 +1,518 @@
+  , isClock    :: IsClock
+  , ty         :: Type
+  , domain     :: Name
+  } deriving (Show, Lift)
+
+data ScaffoldDesc
+  = ScaffoldDesc
+  { functionName  :: Name
+  , functionName# :: Name
+  , functionNameI :: Name
+  , qfunctionName :: Name
+  , qfunctionName# :: Name
+
+  , datatypeNameI :: Name
+  , datatypeNameO :: Name
+
+  , templateName  :: Name
+  , qtemplateName :: Name
+
+  , primitive     :: String
+  } deriving (Show, Lift)
+
+-- * Helper functions for translating between Haskell / Clash / TemplateHaskell
+
+toTemplateHaskellType :: IsClock -> Width -> Name -> Type
+toTemplateHaskellType True _ dom  = AppT (ConT ''C.Clock) (VarT dom)
+toTemplateHaskellType False w dom =
+  AppT
+    (AppT (ConT ''C.Signal) (VarT dom))
+    (AppT (ConT ''C.BitVector) (LitT $ NumTyLit w))
+
+toClashParameter :: Parameter -> HWParameter
+toClashParameter (PString n v) =
+  (Identifier (T.pack n) Nothing, String, Literal Nothing (StringLit v))
+toClashParameter (PInteger n v) =
+  (Identifier (T.pack n) Nothing, Integer, Literal Nothing (NumLit v))
+toClashParameter (PBool n v) =
+  (Identifier (T.pack n) Nothing, Bool, Literal Nothing (BoolLit v))
+toClashParameter (PBitVector n v) =
+  ( Identifier (T.pack n) Nothing
+  , BitVector (fromInteger s)
+  , Literal Nothing (BitVecLit s (toInteger v)))
+ where
+  s = natVal v
+
+toClashType :: ScaffoldPort -> HWType
+toClashType (ScaffoldPort _ _ _ True _ _) = Clock (T.pack "clk")
+toClashType (ScaffoldPort _ w _ False _ _) = BitVector (fromInteger w)
+
+scaffoldPort :: Name -> Port -> ScaffoldPort
+scaffoldPort d (In n w) =
+  ScaffoldPort Netlist.In w n False (toTemplateHaskellType False w d) d
+scaffoldPort d (Out n w) =
+  ScaffoldPort Netlist.Out w n False (toTemplateHaskellType False w d) d
+scaffoldPort d (ClkIn n) =
+  ScaffoldPort Netlist.In 1 n True (toTemplateHaskellType True 1 d) d
+scaffoldPort d (ClkOut n) =
+  ScaffoldPort Netlist.Out 1 n True (toTemplateHaskellType True 1 d) d
+
+scaffoldDomain :: [Port] -> Q (Name, [ScaffoldPort])
+scaffoldDomain ps = do
+  d <- newName "domain"
+  return (d, scaffoldPort d <$> ps)
+
+instantiate
+  :: ScaffoldPort
+  -> Expr
+  -> (Expr, PortDirection, HWType, Expr)
+instantiate p@(ScaffoldPort Netlist.In _ n _ _ _) e
+  = (Identifier (T.pack n) Nothing, Netlist.In, toClashType p, e)
+instantiate p@(ScaffoldPort Netlist.Out _ n _ _ _) e
+  = ( Identifier (T.pack n) Nothing
+    , Netlist.Out, toClashType p, e)
+
+-- *
+
+-- | Build a Clash-Haskell representation of an HDL primitive
+scaffoldTemplate
+  :: Backend s
+  => String
+  -> [Parameter]
+  -> [Name]
+  -> [ScaffoldPort]
+  -> [ScaffoldPort]
+  -> BlackBoxContext
+  -> State s Doc
+scaffoldTemplate primitiveName parameters domains i o bbCtx = do
+  wires <- mapM (ident . T.pack . name) o
+  inst  <- ident (T.pack $ primitiveName <> "_inst")
+  blk   <- ident (T.pack $ primitiveName <> "_blk")
+
+  getMon $ blockDecl blk $ concat
+    [ [ NetDecl Nothing o' t | (o', t) <- zip wires wiresTy ]
```

Comment:
No need for a list comprehension here, you can use `zipWith (NetDecl Nothing) wires wiresTy`